src/QueryPipeline/RemoteQueryExecutorReadContext.cpp

```diff
@@ -147,60 +95,32 @@ bool RemoteQueryExecutorReadContext::checkTimeoutImpl(bool blocking)
 
     if (is_timer_alarmed && !is_socket_ready)
     {
-        /// Socket receive timeout. Drain it in case or error, or it may be hide by timeout exception.
+        /// Socket timeout. Drain it in case of error, or it may be hide by timeout exception.
         timer.drain();
-        throw NetException(ErrorCodes::SOCKET_TIMEOUT, "Timeout exceeded");
+        const String exception_message = getSocketTimeoutExceededMessageByTimeoutType(timeout_type, timeout, connection_fd_description);
+        throw NetException(ErrorCodes::SOCKET_TIMEOUT, exception_message);
     }
 
     return true;
 }
 
-void RemoteQueryExecutorReadContext::setTimer() const
-{
-    /// Did not get packet yet. Init timeout for the next async reading.
-    timer.reset();
-
-    if (receive_timeout_usec)
-        timer.setRelative(receive_timeout_usec);
-}
-
-bool RemoteQueryExecutorReadContext::resumeRoutine()
+void RemoteQueryExecutorReadContext::cancelBefore()
 {
-    if (is_read_in_progress.load(std::memory_order_relaxed) && !checkTimeout())
-        return false;
-
-    {
-        std::lock_guard guard(fiber_lock);
-        if (!fiber)
-            return false;
-
-        fiber = std::move(fiber).resume();
-
-        if (exception)
-            std::rethrow_exception(exception);
-    }
-
-    return true;
-}
-
-void RemoteQueryExecutorReadContext::cancel()
-{
-    std::lock_guard guard(fiber_lock);
-
-    /// It is safe to just destroy fiber - we are not in the process of reading from socket.
-    boost::context::fiber to_destroy = std::move(fiber);
-
     /// One should not try to wait for the current packet here in case of
     /// timeout because this will exceed the timeout.
     /// Anyway if the timeout is exceeded, then the connection will be shutdown
     /// (disconnected), so it will not left in an unsynchronised state.
     if (!is_timer_alarmed)
     {
+        /// If query wasn't sent, just complete sending it.
+        if (!is_query_sent)
+            suspend_when_query_sent = true;
+
         /// Wait for current pending packet, to avoid leaving connection in unsynchronised state.
-        while (is_read_in_progress.load(std::memory_order_relaxed))
+        while (is_in_progress.load(std::memory_order_relaxed))
         {
             checkTimeout(/* blocking= */ true);
-            to_destroy = std::move(to_destroy).resume();
+            resumeUnlocked();
```

Comment:
Now this code works with existing fiber, while previously it had been detached first, though it should not be an issue I guess?

Reply:
It's ok. We call `cancelBefore` under `fiber_lock` mutex in `AsyncTaskExecutor`.